De-dupe .table-dark, add new border color utilities, and update table docs

mdo opened the pull request:
Alternative to #30193. _This ended up being a couple changes in one, hence the length title, but the two are connected._

---

### De-duped .table-dark

Rather than deviate from the `$theme-colors` pattern we have elsewhere, this PR removes the `.table-dark` for changing the background, color, and border of a `.table`. Instead, this can be accomplished through all utilities.

Also removes thead-dark/thead-light modifiers for utilities.

### New border-color utils

To support the utilities-driven approach for replacing `.table-dark`, I've added a few new `border-color` utilities. These are a subset of our available grays, but I do believe they provide an improved set of options without including all the grays (and thus duplicating `$gray-900` and `$dark` options). The new options have been added to the borders utilities docs page.

### New background-color utils

To match the border utilities, I've added one tint lighter for two gray backgrounds (didn't add `$gray-100` as it's the same as `$light`). This way you can choose from additional gray backgrounds and still have a slight contrast in the border colors. Definitely interested in thoughts on this.

### Updated table docs

This also obviously updates the examples in the tables docs to reflect the new dark table approach, but also clarifies how inheritance comes into play with the table styles and the table borders. It also adds a new multiple `<tbody>` example and tweaks a few other bits and pieces.

https://deploy-preview-30342--twbs-bootstrap.netlify.com/docs/4.3/content/tables/

---

Closes #28480, fixes #27879, nullifies #30046, fixes #28160, and fixes #24529.

mdo commented:
For next steps, I'm wondering if we also remove the `.table-*` variants and opt for all background, border, and color variations... Might be pushing it too much.

mdo commented:
Yeah we could perhaps ditch those and create a broader suite of utilities. It was a huge ask awhile back to have light and dark background options. Might be a v6 thing.

MartijnCuppens commented on site/content/docs/4.3/content/tables.md:
```diff
@@ -84,12 +119,10 @@ You can also invert the colors—with light text on dark backgrounds—with `.ta
-### Striped rows
-
-Use `.table-striped` to add zebra-striping to any table row within the `<tbody>`.
+And because this happens with utilities, you have some flexibility in how you style your tables relative to their context. For example, if your page is already dark gray with lighter text, you don't need to set every utility on the `<table>`—just `.border-gray-700`.
```
It took a while before I noticed what was going on here, I was just screening the page without reading the intro text and thought padding was somehow added to the table. I would emphasise the dark page notice more (bold text?) or add what utilities were added to the parent (because you can't see them in the snippet)

MartijnCuppens commented on site/content/docs/4.3/content/tables.md:
```diff
@@ -285,7 +358,43 @@ Add `.table-borderless` for a table without borders.
+And also works with table color classes:
+
+{{< example >}}
+<table class="table table-striped">
```
This table looks pretty ugly 😄

mdo replied:
Yeah I know lol... remove it? Find a better example?

MartijnCuppens commented on site/content/docs/4.3/content/tables.md:
```diff
@@ -575,15 +720,106 @@ Regular table background variants are not available with the dark table, however
 {{< partial "callout-warning-color-assistive-technologies.md" >}}
 {{< /callout >}}
 
+## Nesting
```
This needs more explanation. Also, I'm not a fan of the styles leaking to the nested table. But maybe we should have a look at this afterwards to prevent this PR from becoming too huge?

mdo replied:
Oh yeah, I added the nesting to show how it happens, and it basically convinced me we needed to figure that out. I'm thinking we could find a balance of styling a single first descendant?

Something like this maybe:

```scss
.table {
 > tbody  {
    td,
    th {
      ...
    }
  }
}
```

tmorehouse replied:
> ```scss
>  .table {
>   > tbody  {
>      td,
>      th {
>        ...
>      }
>    }
>  }
> ```

Would still leak into nested table thead and tfoot (as well as tbody)

MartijnCuppens commented on scss/_utilities.scss:
```diff
@@ -77,7 +77,16 @@ $utilities: map-merge(
+        $theme-colors,
+        (
+          "white": $white,
+          "gray-200": $gray-200,
```
The default border color is 300, weird not to have it here:
https://github.com/twbs/bootstrap/blob/8abccff7b3939334c61286df448d0f730da85325/scss/_variables.scss#L318

MartijnCuppens commented:
Wow, the `linear-background()` hack is a genius move!

tmorehouse commented:
The hover and striping background gradient rules need to target the inner `td/th` cells of the body  rows to correctly handle individual cell background colors (and preferably use child selectors to limit leaking into child tables... i.e. preventing hover effects on nested table thead/tfoot) (https://github.com/twbs/bootstrap/issues/29244#issuecomment-595610814)

tmorehouse commented:
One may even want to add in a `table-hover-light` modifier that uses a light translucent gradient when users have dark (almost black) background color.  It would just add two more generated rules and one new SASS variable

tmorehouse commented on scss/_tables.scss:
```diff
@@ -8,11 +8,15 @@
   td {
     padding: $table-cell-padding;
-    border-bottom: $table-border-width solid $table-border-color;
+    // Table cells will only inherit if you use longhand properties
```
Would some CSS minifiers (i.e. cssnano, et al), potentially munge these by trying to merge them into a single shorthand property?

mdo replied:
We use clean-css, which did in fact recombine these. I've disabled that rule for now.

XhmikosR replied:
Maybe we should disable this optimization globally @mdo?

tmorehouse replied:
Maybe there should be a note in the SCSS part of the docs about SASS compilation and minifiers possibly interfering with styles, as some users may be the source SCSS files and a different CSS minifier other than `clean-css` (i.e. `cssnano`, etc)

mdo replied:
Would you mind opening an issue for that @tmorehouse?

tmorehouse replied:
OK...  will do

tmorehouse replied:
Opened #30429

mdo committed:
Drop .table-dark, .thead-dark, and .thead-light for utility classes

mdo committed:
Rewrite table styles to allow inheritance on border-color

CSS doesn't allow inheriting `border-color` on td/th elements when using shorthand notation (e.g., `border: 1px solid #eee`). By separating the values, and disabling clean-css on these rulesets, we allow a single utility class on the `.table` to color our entire table's borders.

mdo committed:
Add new .bg-* and .border-* color utilities

- New background-colors: gray-500, gray-700, and black
- New border-colors: gray-400, gray-600, and black

Note that the border colors are all one shade darker than the backgrounds so that they maintain some contrast.

mdo committed:
Update table docs to reflect all the changes

mdo commented:
Rebased and recombulated the commits here in case we want to pull out the border and background utility changes.

ysds commented:
Can we use the [unset](https://developer.mozilla.org/en-US/docs/Web/CSS/unset) to nested tables if dropping IE11 in v5?

MartijnCuppens commented:
> Can we use the unset to nested tables if dropping IE11 in v5?

We could, but I guess CSS specificity will be a PITA

ysds commented:
> We could, but I guess CSS specificity will be a PITA

Oh I forgot that..

MartijnCuppens commented:
CSS custom properties on the other hand... https://codepen.io/MartijnCuppens/pen/eYNVmzp

@mdo, we probably better switch to a CSS custom properties approach which will help us out with the nesting issues.

tmorehouse commented:
@MartijnCuppens Intereting idea. 

One might have to be careful which the CSS variable names.. perhaps `--table-bg-color`... if similar variable names were used for other things (such as input group background colors, alert background colors, etc) when they are nested inside a table.

MartijnCuppens commented:
@tmorehouse, of course. That was just a quick demo 😉

mdo commented:
So I need to circle back around to this and check in on some things around inheritance and nesting. I don't think mixing the CSS vars and Sass makes sense necessarily? Maybe I'm missing something there.

MartijnCuppens commented:
I'm working on an alternative solution, will open another PR soon.

stgm commented:
Has the use of linear-gradient been performance-tested in some way? I have an example where this makes Safari rendering extremely slow when scrolling. Removing the gradient helps, but obviously kills my striping. If this wasn't expected, I can have a look at creating an example and posting an issue.

MartijnCuppens commented:
> If this wasn't expected, I can have a look at creating an example and posting an issue.

Please do if there are any performance issue
